File: global/src/main/java/ms2709/global/SelfValidating.java
package ms2709.global;


import jakarta.validation.*;
import java.util.Set;

/**
 *  Command객체의 유효성 검증을 위한 추상 클래스
 *
 * @author 심문섭
 * @version 1.0
 * @class SelfValidating
 * @since 2024-05-02 오전 8:24
 */
public abstract class SelfValidating<T> {

    private Validator validator;

    protected SelfValidating() {
        ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
        validator = factory.getValidator();
    }

    /**
     * Evaluates all Bean Validations on the attributes of this
     * instance.
     */
    protected void validateSelf() {
        Set<ConstraintViolation<T>> violations = validator.validate((T) this);
        if (!violations.isEmpty()) {
            throw new ConstraintViolationException(violations);
        }
    }
}

File: member-service/src/main/kotlin/ms2709/member/global/SelfValidating.kt
package ms2709.member.global

import jakarta.validation.ConstraintViolationException
import jakarta.validation.Validation
import jakarta.validation.Validator

/**
 *
 * Command객체의 유효성 검증을 위한 추상 클래스
 *
 * @class SelfValidating
 * @author 심문섭
 * @version 1.0
 * @since 2024-04-29 9:51PM
 */
abstract class SelfValidating<T> {
    private val validator: Validator
    constructor() {
        Validation.buildDefaultValidatorFactory().run {
            this@SelfValidating.validator = this.validator
        }
    }

    protected fun validateSelf() {
        val violations = validator.validate(this)
        if (violations.isNotEmpty()) {
            throw ConstraintViolationException(violations)
        }
    }
}